tools/syz-trace2syz/proggen/call_selector.go

```diff
@@ -22,40 +25,82 @@ var discriminatorArgs = map[string][]int{
 	"sendto":      {0},
 	"sendmsg":     {0},
 	"getsockname": {0},
+	"openat":      {1},
 }
 
-type callSelector struct {
+var openDiscriminatorArgs = map[string]int{
+	"open":         0,
+	"openat":       1,
+	"syz_open_dev": 0,
+}
+
+type callSelector interface {
+	Select(call *parser.Syscall) *prog.Syscall
+}
+
+func newSelectors(target *prog.Target, returnCache returnCache) []callSelector {
+	sc := newSelectorCommon(target, returnCache)
+	return []callSelector{
+		&defaultCallSelector{sc},
+		&openCallSelector{sc},
+	}
+}
+
+type selectorCommon struct {
 	target      *prog.Target
 	returnCache returnCache
 	callCache   map[string][]*prog.Syscall
 }
 
-func newCallSelector(target *prog.Target, returnCache returnCache) *callSelector {
-	return &callSelector{
+func newSelectorCommon(target *prog.Target, returnCache returnCache) *selectorCommon {
+	return &selectorCommon{
 		target:      target,
 		returnCache: returnCache,
 		callCache:   make(map[string][]*prog.Syscall),
 	}
 }
 
-// Select returns the best matching descrimination for this syscall.
-func (cs *callSelector) Select(call *parser.Syscall) *prog.Syscall {
-	match := cs.target.SyscallMap[call.CallName]
-	discriminators := discriminatorArgs[call.CallName]
-	if len(discriminators) == 0 {
-		return match
-	}
-	score := 0
-	for _, meta := range cs.callSet(call.CallName) {
-		if score1 := cs.matchCall(meta, call, discriminators); score1 > score {
-			match, score = meta, score1
+// matches strace file string with a constant string in openat or syz_open_dev
+// if the string in openat or syz_open_dev has a # then this method will
+// return the corresponding  id from the strace string
+func (cs *selectorCommon) matchFilename(syzFile, straceFile []byte) (bool, int) {
+	syzFile = bytes.Trim(syzFile, "\x00")
+	straceFile = bytes.Trim(straceFile, "\x00")
+	// we should have exact matches except for cases where syzFile has a #
+	i, j := 0, 0
+	var id []byte
+	dev := -1
+	for j < len(straceFile) && i < len(syzFile) {
+		if syzFile[i] == straceFile[j] {
+			i++
+			j++
+			continue
 		}
+		// if there is a character mismatch then it has to be a #
+		if syzFile[i] != '#' {
+			return false, -1
+		}
+		i++
+		// the corresponding mismatched character in straceFile
+		// needs to be a digit.
+		if !unicode.IsDigit(rune(straceFile[j])) {
+			return false, -1
+		}
+		for ; j < len(straceFile) &&
+			unicode.IsDigit(rune(straceFile[j])); j++ {
```

Comment:
We don't substitute `#` with more than 1 char.
Also, multiple replacements for `#` should be concatenated, while this code takes only the last one.
E.g. for existing test:
```
"/dev/some#/some#", "/dev/some01/some01", 1, true,
```
we should return 11 (dec).

I think we need something along the following lines:
```
func (cs *selectorCommon) matchFilename(syzFile, straceFile []byte) (bool, int) {
	syzFile = bytes.Trim(syzFile, "\x00")
	straceFile = bytes.Trim(straceFile, "\x00")
	if len(syzFile) != len(straceFile) {
		return false, -1
	}
	var id []byte
	for i, c := range syzFile {
		x := straceFile[i]
		if c == x {
			continue
		}
		if c != '#' || !unicode.IsDigit(rune(x)) {
			return false, -1
		}
		id = append(id, x)
	}
	dev, _ := strconv.Atoi(string(id))
	return true, dev
}
```

Reply:
thanks! done